File: Traditional-Important-Algorithms/exercises_py/09_rate_limiting/rate07_bulkhead.py
# I AM NOT DONE

"""
Exercise: Bulkhead Isolation Pattern

Bulkhead is a fault tolerance pattern that isolates resources into pools
to prevent failure in one area from affecting others.

Inspired by ship bulkheads that prevent entire ship from sinking if one
compartment is breached.

How it works:
- Divide resources into isolated pools (e.g., thread pools per service)
- Each pool has fixed capacity
- Failure/exhaustion in one pool doesn't affect others
- Limits blast radius of failures

Your task: Implement a bulkhead isolation pattern.

Key concepts:
- Resource pool isolation
- Capacity limits per bulkhead
- Independent failure domains
- Resource acquisition and release
"""

from typing import Callable, TypeVar, Dict, List, Optional
from dataclasses import dataclass


T = TypeVar('T')


@dataclass
class BulkheadId:
    """Identifier for a bulkhead"""
    id: int

    def __hash__(self):
        return hash(self.id)

    def __eq__(self, other):
        if isinstance(other, BulkheadId):
            return self.id == other.id
        return False


@dataclass
class BulkheadStats:
    """Statistics for a bulkhead"""
    name: str
    capacity: int
    in_use: int
    available: int
    total_accepted: int
    total_rejected: int
    utilization: float  # in_use / capacity


class BulkheadPool:
    """Internal representation of a bulkhead pool"""

    def __init__(self, name: str, capacity: int):
        self.name = name
        self.capacity = capacity
        self.in_use = 0
        self.total_accepted = 0
        self.total_rejected = 0


class BulkheadIsolation:
    """Bulkhead isolation system for resource management"""

    def __init__(self):
        """Initialize empty bulkhead isolation system"""
        # TODO: Initialize empty bulkhead isolation system
        pass

    def create_bulkhead(self, bulkhead_id: BulkheadId, name: str, capacity: int):
        """Create a new bulkhead with given parameters"""
        # TODO: Create a new bulkhead with given parameters
        # - Initialize BulkheadPool with capacity and name
        # - Set in_use, total_accepted, total_rejected to 0
        # - Insert into bulkheads map
        pass

    def try_acquire(self, bulkhead_id: BulkheadId) -> bool:
        """
        Try to acquire a resource slot from bulkhead

        Args:
            bulkhead_id: ID of bulkhead to acquire from

        Returns:
            True if acquired, False if not found or full

        Raises:
            BulkheadError: If bulkhead not found or full
        """
        # TODO: Try to acquire a resource slot from bulkhead
        # - Check if bulkhead exists, raise BulkheadError("Not found") if not
        # - Check if in_use < capacity
        # - If yes, increment in_use and total_accepted, return True
        # - If no, increment total_rejected, raise BulkheadError("Full")
        pass

    def release(self, bulkhead_id: BulkheadId):
        """
        Release a resource slot back to bulkhead

        Args:
            bulkhead_id: ID of bulkhead to release to

        Raises:
            BulkheadError: If bulkhead not found
        """
        # TODO: Release a resource slot back to bulkhead
        # - Check if bulkhead exists
        # - Decrement in_use (but don't go below 0)
        # - Raise BulkheadError if not found
        pass

    def execute(self, bulkhead_id: BulkheadId, operation: Callable[[], T]) -> T:
        """
        Execute operation with bulkhead protection

        Args:
            bulkhead_id: ID of bulkhead to use
            operation: Function to execute

        Returns:
            Result of operation

        Raises:
            BulkheadError: If bulkhead issues occur
        """
        # TODO: Execute operation with bulkhead protection
        # - Try to acquire slot
        # - If successful, execute operation
        # - Release slot (even if operation raises - use try/finally)
        # - Return result or propagate BulkheadError
        pass

    def get_stats(self, bulkhead_id: BulkheadId) -> Optional[BulkheadStats]:
        """Return statistics for given bulkhead"""
        # TODO: Return statistics for given bulkhead
        # - Return None if not found
        # - Return BulkheadStats with current values
        pass

    def available_capacity(self, bulkhead_id: BulkheadId) -> Optional[int]:
        """Return available capacity for bulkhead"""
        # TODO: Return available capacity for bulkhead
        # - Return None if not found
        # - Return capacity - in_use
        pass

    def is_full(self, bulkhead_id: BulkheadId) -> bool:
        """Check if bulkhead is at capacity"""
        # TODO: Check if bulkhead is at capacity
        # - Return True if in_use >= capacity
        # - Return False if not found or has capacity
        pass

    def reset(self, bulkhead_id: BulkheadId):
        """
        Reset bulkhead statistics (keep capacity)

        Args:
            bulkhead_id: ID of bulkhead to reset

        Raises:
            BulkheadError: If bulkhead not found
        """
        # TODO: Reset bulkhead statistics (keep capacity)
        # - Reset in_use, total_accepted, total_rejected to 0
        pass

    def list_bulkheads(self) -> List[BulkheadId]:
        """Return list of all bulkhead IDs"""
        # TODO: Return list of all bulkhead IDs
        pass


class BulkheadError(Exception):
    """Exception raised for bulkhead errors"""
    pass


import unittest


class TestBulkheadIsolation(unittest.TestCase):
    def test_create_bulkhead(self):
        isolation = BulkheadIsolation()
        isolation.create_bulkhead(BulkheadId(1), "service-a", 5)

        stats = isolation.get_stats(BulkheadId(1))
        self.assertIsNotNone(stats)
        self.assertEqual(stats.name, "service-a")
        self.assertEqual(stats.capacity, 5)
        self.assertEqual(stats.in_use, 0)

    def test_acquire_and_release(self):
        isolation = BulkheadIsolation()
        isolation.create_bulkhead(BulkheadId(1), "test", 3)

        self.assertTrue(isolation.try_acquire(BulkheadId(1)))
        self.assertEqual(isolation.available_capacity(BulkheadId(1)), 2)

        isolation.release(BulkheadId(1))
        self.assertEqual(isolation.available_capacity(BulkheadId(1)), 3)

    def test_capacity_limit(self):
        isolation = BulkheadIsolation()
        isolation.create_bulkhead(BulkheadId(1), "limited", 2)

        self.assertTrue(isolation.try_acquire(BulkheadId(1)))
        self.assertTrue(isolation.try_acquire(BulkheadId(1)))

        with self.assertRaises(BulkheadError):
            isolation.try_acquire(BulkheadId(1))

        self.assertTrue(isolation.is_full(BulkheadId(1)))

    def test_nonexistent_bulkhead(self):
        isolation = BulkheadIsolation()

        with self.assertRaises(BulkheadError):
            isolation.try_acquire(BulkheadId(99))

        self.assertIsNone(isolation.get_stats(BulkheadId(99)))

    def test_execute_with_bulkhead(self):
        isolation = BulkheadIsolation()
        isolation.create_bulkhead(BulkheadId(1), "exec", 3)

        result = isolation.execute(BulkheadId(1), lambda: 42)
        self.assertEqual(result, 42)

        # After execution, slot should be released
        self.assertEqual(isolation.available_capacity(BulkheadId(1)), 3)

    def test_execute_when_full(self):
        isolation = BulkheadIsolation()
        isolation.create_bulkhead(BulkheadId(1), "full", 1)

        isolation.try_acquire(BulkheadId(1))

        with self.assertRaises(BulkheadError):
            isolation.execute(BulkheadId(1), lambda: 42)

    def test_isolation_between_bulkheads(self):
        isolation = BulkheadIsolation()
        isolation.create_bulkhead(BulkheadId(1), "service-a", 2)
        isolation.create_bulkhead(BulkheadId(2), "service-b", 2)

        # Fill bulkhead 1
        isolation.try_acquire(BulkheadId(1))
        isolation.try_acquire(BulkheadId(1))

        # Bulkhead 1 full, but bulkhead 2 still available
        self.assertTrue(isolation.is_full(BulkheadId(1)))
        self.assertFalse(isolation.is_full(BulkheadId(2)))
        self.assertTrue(isolation.try_acquire(BulkheadId(2)))

    def test_statistics_tracking(self):
        isolation = BulkheadIsolation()
        isolation.create_bulkhead(BulkheadId(1), "stats", 2)

        isolation.try_acquire(BulkheadId(1))
        isolation.try_acquire(BulkheadId(1))
        try:
            isolation.try_acquire(BulkheadId(1))  # Rejected
        except:
            pass

        stats = isolation.get_stats(BulkheadId(1))
        self.assertEqual(stats.total_accepted, 2)
        self.assertEqual(stats.total_rejected, 1)

    def test_utilization_calculation(self):
        isolation = BulkheadIsolation()
        isolation.create_bulkhead(BulkheadId(1), "util", 4)

        isolation.try_acquire(BulkheadId(1))
        isolation.try_acquire(BulkheadId(1))

        stats = isolation.get_stats(BulkheadId(1))
        self.assertAlmostEqual(stats.utilization, 0.5, places=2)  # 2/4 = 50%

    def test_reset(self):
        isolation = BulkheadIsolation()
        isolation.create_bulkhead(BulkheadId(1), "reset", 3)

        isolation.try_acquire(BulkheadId(1))
        isolation.try_acquire(BulkheadId(1))

        isolation.reset(BulkheadId(1))

        stats = isolation.get_stats(BulkheadId(1))
        self.assertEqual(stats.in_use, 0)
        self.assertEqual(stats.total_accepted, 0)
        self.assertEqual(stats.total_rejected, 0)
        self.assertEqual(stats.capacity, 3)  # Capacity unchanged

    def test_list_bulkheads(self):
        isolation = BulkheadIsolation()
        isolation.create_bulkhead(BulkheadId(1), "a", 1)
        isolation.create_bulkhead(BulkheadId(2), "b", 2)
        isolation.create_bulkhead(BulkheadId(3), "c", 3)

        ids = isolation.list_bulkheads()
        id_values = sorted([bid.id for bid in ids])

        self.assertEqual(id_values, [1, 2, 3])

    def test_multiple_operations(self):
        isolation = BulkheadIsolation()
        isolation.create_bulkhead(BulkheadId(1), "multi", 3)

        r1 = isolation.execute(BulkheadId(1), lambda: "first")
        r2 = isolation.execute(BulkheadId(1), lambda: "second")
        r3 = isolation.execute(BulkheadId(1), lambda: "third")

        self.assertEqual(r1, "first")
        self.assertEqual(r2, "second")
        self.assertEqual(r3, "third")

        stats = isolation.get_stats(BulkheadId(1))
        self.assertEqual(stats.total_accepted, 3)
        self.assertEqual(stats.in_use, 0)  # All released


if __name__ == '__main__':
    unittest.main()

File: Traditional-Important-Algorithms/exercises/09_rate_limiting/rate07_bulkhead.rs
// rate07_bulkhead.rs
//
// Bulkhead is a fault tolerance pattern that isolates resources into pools
// to prevent failure in one area from affecting others.
//
// Inspired by ship bulkheads that prevent entire ship from sinking if one
// compartment is breached.
//
// How it works:
// - Divide resources into isolated pools (e.g., thread pools per service)
// - Each pool has fixed capacity
// - Failure/exhaustion in one pool doesn't affect others
// - Limits blast radius of failures
//
// Your task: Implement a bulkhead isolation pattern.
//
// Key concepts:
// - Resource pool isolation
// - Capacity limits per bulkhead
// - Independent failure domains
// - Resource acquisition and release

// I AM NOT DONE

use std::collections::HashMap;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct BulkheadId(pub usize);

#[derive(Debug)]
struct BulkheadPool {
    name: String,
    capacity: usize,
    in_use: usize,
    total_accepted: usize,
    total_rejected: usize,
}

pub struct BulkheadIsolation {
    bulkheads: HashMap<BulkheadId, BulkheadPool>,
}

impl BulkheadIsolation {
    pub fn new() -> Self {
        // TODO: Initialize empty bulkhead isolation system
        todo!()
    }

    pub fn create_bulkhead(&mut self, id: BulkheadId, name: String, capacity: usize) {
        // TODO: Create a new bulkhead with given parameters
        // - Initialize BulkheadPool with capacity and name
        // - Set in_use, total_accepted, total_rejected to 0
        // - Insert into bulkheads map
        todo!()
    }

    pub fn try_acquire(&mut self, id: BulkheadId) -> Result<(), BulkheadError> {
        // TODO: Try to acquire a resource slot from bulkhead
        // - Check if bulkhead exists, return Err(BulkheadError::NotFound) if not
        // - Check if in_use < capacity
        // - If yes, increment in_use and total_accepted, return Ok(())
        // - If no, increment total_rejected, return Err(BulkheadError::Full)
        todo!()
    }

    pub fn release(&mut self, id: BulkheadId) -> Result<(), BulkheadError> {
        // TODO: Release a resource slot back to bulkhead
        // - Check if bulkhead exists
        // - Decrement in_use (but don't go below 0)
        // - Return Ok(()) or Err if not found
        todo!()
    }

    pub fn execute<F, T>(&mut self, id: BulkheadId, operation: F) -> Result<T, BulkheadError>
    where
        F: FnOnce() -> T,
    {
        // TODO: Execute operation with bulkhead protection
        // - Try to acquire slot
        // - If successful, execute operation
        // - Release slot (even if operation panics - use a guard or careful error handling)
        // - Return result wrapped in Ok, or propagate BulkheadError
        todo!()
    }

    pub fn get_stats(&self, id: BulkheadId) -> Option<BulkheadStats> {
        // TODO: Return statistics for given bulkhead
        // - Return None if not found
        // - Return Some(BulkheadStats) with current values
        todo!()
    }

    pub fn available_capacity(&self, id: BulkheadId) -> Option<usize> {
        // TODO: Return available capacity for bulkhead
        // - Return None if not found
        // - Return Some(capacity - in_use)
        todo!()
    }

    pub fn is_full(&self, id: BulkheadId) -> bool {
        // TODO: Check if bulkhead is at capacity
        // - Return true if in_use >= capacity
        // - Return false if not found or has capacity
        todo!()
    }

    pub fn reset(&mut self, id: BulkheadId) -> Result<(), BulkheadError> {
        // TODO: Reset bulkhead statistics (keep capacity)
        // - Reset in_use, total_accepted, total_rejected to 0
        todo!()
    }

    pub fn list_bulkheads(&self) -> Vec<BulkheadId> {
        // TODO: Return list of all bulkhead IDs
        todo!()
    }
}

impl Default for BulkheadIsolation {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum BulkheadError {
    NotFound,
    Full,
}

#[derive(Debug, Clone, PartialEq)]
pub struct BulkheadStats {
    pub name: String,
    pub capacity: usize,
    pub in_use: usize,
    pub available: usize,
    pub total_accepted: usize,
    pub total_rejected: usize,
    pub utilization: f64, // in_use / capacity
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_create_bulkhead() {
        let mut isolation = BulkheadIsolation::new();
        isolation.create_bulkhead(BulkheadId(1), "service-a".to_string(), 5);

        let stats = isolation.get_stats(BulkheadId(1)).unwrap();
        assert_eq!(stats.name, "service-a");
        assert_eq!(stats.capacity, 5);
        assert_eq!(stats.in_use, 0);
    }

    #[test]
    fn test_acquire_and_release() {
        let mut isolation = BulkheadIsolation::new();
        isolation.create_bulkhead(BulkheadId(1), "test".to_string(), 3);

        assert!(isolation.try_acquire(BulkheadId(1)).is_ok());
        assert_eq!(isolation.available_capacity(BulkheadId(1)), Some(2));

        assert!(isolation.release(BulkheadId(1)).is_ok());
        assert_eq!(isolation.available_capacity(BulkheadId(1)), Some(3));
    }

    #[test]
    fn test_capacity_limit() {
        let mut isolation = BulkheadIsolation::new();
        isolation.create_bulkhead(BulkheadId(1), "limited".to_string(), 2);

        assert!(isolation.try_acquire(BulkheadId(1)).is_ok());
        assert!(isolation.try_acquire(BulkheadId(1)).is_ok());
        assert_eq!(isolation.try_acquire(BulkheadId(1)), Err(BulkheadError::Full));

        assert!(isolation.is_full(BulkheadId(1)));
    }

    #[test]
    fn test_nonexistent_bulkhead() {
        let mut isolation = BulkheadIsolation::new();

        assert_eq!(isolation.try_acquire(BulkheadId(99)), Err(BulkheadError::NotFound));
        assert_eq!(isolation.get_stats(BulkheadId(99)), None);
    }

    #[test]
    fn test_execute_with_bulkhead() {
        let mut isolation = BulkheadIsolation::new();
        isolation.create_bulkhead(BulkheadId(1), "exec".to_string(), 3);

        let result = isolation.execute(BulkheadId(1), || 42);
        assert_eq!(result, Ok(42));

        // After execution, slot should be released
        assert_eq!(isolation.available_capacity(BulkheadId(1)), Some(3));
    }

    #[test]
    fn test_execute_when_full() {
        let mut isolation = BulkheadIsolation::new();
        isolation.create_bulkhead(BulkheadId(1), "full".to_string(), 1);

        isolation.try_acquire(BulkheadId(1)).ok();

        let result = isolation.execute(BulkheadId(1), || 42);
        assert_eq!(result, Err(BulkheadError::Full));
    }

    #[test]
    fn test_isolation_between_bulkheads() {
        let mut isolation = BulkheadIsolation::new();
        isolation.create_bulkhead(BulkheadId(1), "service-a".to_string(), 2);
        isolation.create_bulkhead(BulkheadId(2), "service-b".to_string(), 2);

        // Fill bulkhead 1
        isolation.try_acquire(BulkheadId(1)).ok();
        isolation.try_acquire(BulkheadId(1)).ok();

        // Bulkhead 1 full, but bulkhead 2 still available
        assert!(isolation.is_full(BulkheadId(1)));
        assert!(!isolation.is_full(BulkheadId(2)));
        assert!(isolation.try_acquire(BulkheadId(2)).is_ok());
    }

    #[test]
    fn test_statistics_tracking() {
        let mut isolation = BulkheadIsolation::new();
        isolation.create_bulkhead(BulkheadId(1), "stats".to_string(), 2);

        isolation.try_acquire(BulkheadId(1)).ok();
        isolation.try_acquire(BulkheadId(1)).ok();
        isolation.try_acquire(BulkheadId(1)).ok(); // Rejected

        let stats = isolation.get_stats(BulkheadId(1)).unwrap();
        assert_eq!(stats.total_accepted, 2);
        assert_eq!(stats.total_rejected, 1);
    }

    #[test]
    fn test_utilization_calculation() {
        let mut isolation = BulkheadIsolation::new();
        isolation.create_bulkhead(BulkheadId(1), "util".to_string(), 4);

        isolation.try_acquire(BulkheadId(1)).ok();
        isolation.try_acquire(BulkheadId(1)).ok();

        let stats = isolation.get_stats(BulkheadId(1)).unwrap();
        assert!((stats.utilization - 0.5).abs() < 0.01); // 2/4 = 50%
    }

    #[test]
    fn test_reset() {
        let mut isolation = BulkheadIsolation::new();
        isolation.create_bulkhead(BulkheadId(1), "reset".to_string(), 3);

        isolation.try_acquire(BulkheadId(1)).ok();
        isolation.try_acquire(BulkheadId(1)).ok();

        isolation.reset(BulkheadId(1)).ok();

        let stats = isolation.get_stats(BulkheadId(1)).unwrap();
        assert_eq!(stats.in_use, 0);
        assert_eq!(stats.total_accepted, 0);
        assert_eq!(stats.total_rejected, 0);
        assert_eq!(stats.capacity, 3); // Capacity unchanged
    }

    #[test]
    fn test_list_bulkheads() {
        let mut isolation = BulkheadIsolation::new();
        isolation.create_bulkhead(BulkheadId(1), "a".to_string(), 1);
        isolation.create_bulkhead(BulkheadId(2), "b".to_string(), 2);
        isolation.create_bulkhead(BulkheadId(3), "c".to_string(), 3);

        let mut ids = isolation.list_bulkheads();
        ids.sort_by_key(|id| id.0);

        assert_eq!(ids, vec![BulkheadId(1), BulkheadId(2), BulkheadId(3)]);
    }

    #[test]
    fn test_multiple_operations() {
        let mut isolation = BulkheadIsolation::new();
        isolation.create_bulkhead(BulkheadId(1), "multi".to_string(), 3);

        let r1 = isolation.execute(BulkheadId(1), || "first");
        let r2 = isolation.execute(BulkheadId(1), || "second");
        let r3 = isolation.execute(BulkheadId(1), || "third");

        assert_eq!(r1, Ok("first"));
        assert_eq!(r2, Ok("second"));
        assert_eq!(r3, Ok("third"));

        let stats = isolation.get_stats(BulkheadId(1)).unwrap();
        assert_eq!(stats.total_accepted, 3);
        assert_eq!(stats.in_use, 0); // All released
    }
}
